PhysicsTools/PatUtils/plugins/L1PrefiringWeightProducer.cc

```diff
@@ -0,0 +1,469 @@
+        nonPrefiringProba[fluct] *= (1. - prefiringprob_mu);
+        nonPrefiringProbaMuon[fluct] *= (1. - prefiringprob_mu);
+      }
+    }
+  }
+  for (const auto fluct :
+       {fluctuations::upSyst, fluctuations::downSyst, fluctuations::upStat, fluctuations::downStat}) {
+    if (!missingInputMuon_) {
+      for (const auto& muon : theMuons) {
+        double pt = muon.pt();
+        double phi = muon.phi();
+        double eta = muon.eta();
+        // Remove crappy tracker muons which would not have prefired the L1 trigger
+        if (pt < 5 && !muon.isStandAloneMuon())
+          continue;
+        double prefiringprob_mu = getPrefiringRateMuon(eta, phi, pt, fluct);
+        nonPrefiringProbaMuon[fluct] *= (1. - prefiringprob_mu);
+      }
+    }
+  }
+  //Move global prefire weights, as well as those for muons, photons, and jets, to the event
+  iEvent.emplace(nonPrefiringProbToken_, nonPrefiringProba[0]);
+  iEvent.emplace(nonPrefiringProbUpToken_, nonPrefiringProba[1]);
+  iEvent.emplace(nonPrefiringProbDownToken_, nonPrefiringProba[2]);
+
+  iEvent.emplace(nonPrefiringProbJetToken_, nonPrefiringProbaJet[0]);
+  iEvent.emplace(nonPrefiringProbJetUpToken_, nonPrefiringProbaJet[1]);
+  iEvent.emplace(nonPrefiringProbJetDownToken_, nonPrefiringProbaJet[2]);
+
+  iEvent.emplace(nonPrefiringProbPhotonToken_, nonPrefiringProbaPhoton[0]);
+  iEvent.emplace(nonPrefiringProbPhotonUpToken_, nonPrefiringProbaPhoton[1]);
+  iEvent.emplace(nonPrefiringProbPhotonDownToken_, nonPrefiringProbaPhoton[2]);
+
+  iEvent.emplace(nonPrefiringProbMuonToken_, nonPrefiringProbaMuon[0]);
+  iEvent.emplace(nonPrefiringProbMuonUpSystToken_, nonPrefiringProbaMuon[3]);
+  iEvent.emplace(nonPrefiringProbMuonDownSystToken_, nonPrefiringProbaMuon[4]);
+  iEvent.emplace(nonPrefiringProbMuonUpStatToken_, nonPrefiringProbaMuon[5]);
+  iEvent.emplace(nonPrefiringProbMuonDownStatToken_, nonPrefiringProbaMuon[6]);
+}
+
+double L1PrefiringWeightProducer::getPrefiringRateEcal(double eta,
+                                                       double pt,
+                                                       TH2F* h_prefmap,
+                                                       fluctuations fluctuation) const {
+  //Check pt is not above map overflow
+  int nbinsy = h_prefmap->GetNbinsY();
+  double maxy = h_prefmap->GetYaxis()->GetBinLowEdge(nbinsy + 1);
+  if (pt >= maxy)
+    pt = maxy - 0.01;
+  int thebin = h_prefmap->FindBin(eta, pt);
+
+  double prefrate = h_prefmap->GetBinContent(thebin);
+
+  double statuncty = h_prefmap->GetBinError(thebin);
+  double systuncty = prefiringRateSystUncEcal_ * prefrate;
+
+  if (fluctuation == up)
+    prefrate = std::min(1., prefrate + sqrt(pow(statuncty, 2) + pow(systuncty, 2)));
+  else if (fluctuation == down)
+    prefrate = std::max(0., prefrate - sqrt(pow(statuncty, 2) + pow(systuncty, 2)));
+  return prefrate;
+}
+double L1PrefiringWeightProducer::getPrefiringRateMuon(double eta,
```

Comment:
Please add one empty line, to visually separate the two methods in the code

Reply:
Done.